### PR TITLE
Use CRI to pull images

### PR DESCRIPTION
**Pull request by ecordell**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This is an initial attempt to use CRI to pull the images to mount. This should make it simpler to configure image volumes, because registry configuration (including mirrors, proxies, auth, etc) will come from the node.

**Special notes for your reviewer**:

This is currently a prototype, and shouldn't be used (or merged)

### CODE REVIEW

File: pkg/image/nodeserver.go [hunk 1]
```diff
@@ -148,53 +143,95 @@ func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
...
 
-	timeout := false
-	if ns.Timeout > 0 {
-		timer := time.AfterFunc(ns.Timeout, func() {
-			timeout = true
-			// TODO: cmd.Stop()
-		})
-		defer timer.Stop()
+	// load in a statically built `cp` that can copy content from within the container
+	// this is necessary because there is no way to get a hostPath for a container's fs via CRI
```
Comment (glennpratt) [inline pkg/image/nodeserver.go:216]: It seems so close though:

```
root@microk8s:/home/ubuntu# export CONTAINER_RUNTIME_ENDPOINT=unix:///var/snap/microk8s/common/run/containerd.sock
root@microk8s:/home/ubuntu# crictl info | jq .config.stateDir
"/var/snap/microk8s/common/run/containerd/io.containerd.grpc.v1.cri"
root@microk8s:/home/ubuntu# ls /var/snap/microk8s/common/run/containerd/io.containerd.grpc.v1.cri/sandboxes/
01ae0f920ddd24f57b3babd0dfb0391572ac4806802b6fde238bfc36e7c8f522  9dbc397314f3444aa2a7a191a3c26f033817068dd796d88f29d57b44c63a75a1
1183a30567077617fdf2e81fe88527dabb86f779e1e3c59f5d1199e5b4856ac4  a1d6e737925a4659b5d77c6fdfed7338c8d296330ab58f47e85908660d8d3064
11f7ff2640065576257a9f508547c0936b3cd2df14a170ef19e83c5f5bd6c2b8  b9fd3cb11b07b557c3143fe7c556a2289c7b0acb851122aba7bb03093e7b65ad
24bf1a317ad4c6711ff09e42dda8a7635614c186b61d7a88a5dfe7c1488ded8f  c0db472c326dc09c47976c9f6e21cd5174e96e7c091916c9e72922aed40271d0
254c98f107665d92b09467ce635371a3e366f4e6bb1aba8cfdfa6688c182be66  c8aac2fbc4fb5c618f22fe0050cda81fb3ee0b4330c953fb6abd8831e90f4b2e
36e774f6ab628031c8ba38b9bf01d2906e68ba641e45edc2e7e4a4aab11f0484  c994f69c120d7c8603e9d9806bc2e9d9b66bc2231eb738926f1e19948b19fd66
388743b8f01edd4f11495ebada3a37c5d63a9f74972d0a8410600f17cd6650dc  ca23c10562ae28b895d9f95f56d9cfb92754a9480932a2fff61c11d796f63a07
39bfc2f1d4b69578948b51ae9a85a1ea65454a73616ad2bf3971bc5bdb9dd517  cc704c0ca7181f92b92de751aea86ca0e0ec852f5784b80d75931a0ed9bea5a4
483b1c3c2a799360d1e53e148a022231af17b89b4cfab514d3a88f1b13417464  d295c987e8e8b9ae8ae61bb33b901d2db56248b00d41576ee68855a6abcd5a53
509d256a14fd076be7f660b7ca6b0cd8a0a4a84c427fc7cea3654792047d03bf  d36dc90d4626bbe1633d4f34baae06ab76298de1ef5ffa25b46510cc710e7dc9
544776f72ff33918030950124f9c39907e001c4b278da11d8ba04871ac2283e7  df30a514624761c142dbaa96c578a2b49637e01c8a512f471a48d87e6aade5be
5ff81f47f9302c2d62dc3c9d5dce72f40b5cd31a49f7003f00cb7b668dec4756  ee4ddb72fc69bc6d077e602d5134bcd8932a311b2a7c260501f0105c0a8b90ef
7344d19e5ebe9cd2467acc8a98cab476d2980501941e783d207d779281628377  f096b1f4f584044994896d913de2fe57235225c3779acdfaf86d9f75daa8ca9d
7433078867166e0ef2409d3751bec9401c51ecd23d9709375bc6eb0a1dd27a0c  f19a0d680638b0bf7849e6055877ef9bfa24fa409a860643aac605a979cb1a4d
7b1d6f10a39b76ac720650ef57e59a2bef83bc52cc7216fbd92c216b329f97f4  f21813d35f844575baf7be3e523e7c14fc510a98f38eb07e2c1f2e180ee73fe0
7ee029b1d2d94b4705f6aed7a7da87fdb7edffe25cde631518715e1d6070ea3d  ff93e9c87ac9f1cdac431c08368b0a60cfc6aee49e88982de4fff453cc3f5544
92412bbf8099869eb9465a4ed11fa7bc5f7cf849b75671eb843c94e22534becd
```

**Comment by kfox1111** [comment]
I'd still really like to find a way to avoid duplicating the data like the existing implementation does.

**Comment by rhatdan** [comment]
We (@vrothberg and me) have added functionality to Podman to allow the mounting of images on the host, as well as mounting images into a podman container.  
mnt:=$(podman image mount fedora)
Then $mnt could be used as the source for a CRI-O container image.

Is this something we should work on together?

**Comment by kfox1111** [comment]
Interesting. Does that do the equivalent of buildah mount? The code in trunk currently is just doing a buildah mount then passing it back to kubelet to mount into the other container.

**Comment by ecordell** [comment]
@rhatdan the goal of this PR was to enable a csd-driver-image-populator that would be CRI-implementation independent. Using podman would have similar issues to using buildah (registry configuration is cri-implementation specific, so credential management for volume images is harder)

**Comment by glennpratt** [comment]
Has anyone reviewed this project? https://github.com/warm-metal/csi-driver-image

**Comment by kfox1111** [comment]
No, hadn't seen that. interesting. Looks like it only works with docker/containerd so far?

**Pull request closed without merging**